### PR TITLE
added fs-proxy and moved setup to package.json closes Devcord/cordlr#55 closes Devcord/cordlr#67

### PR DESCRIPTION
**Pull request by mees-**
As i  mentioned in #67 i added in fs-proxy, the plugins should now be able to write to the config file, an optional restriction could be to create a property in the config for every plugins and only give them (read) access to that property

### CODE REVIEW

File: bin/cordlr [hunk 1]
```diff
@@ -20,17 +15,17 @@ main()
 
 function main () {
   // --help, -h, help
-  if (flags.help || flags.h || action === 'help') return help()
+  if (flags.help || flags.h || action === 'help') return require('./methods/help')()
```
Comment (jamen) [inline bin/cordlr]: Hmm, I don't agree with these changes.  It's not _too_ big of a deal to me, but I don't really think it adds any benefit changing it from the way it was to `require`.

I don't think saving 5 lines up there warrants making the code less readable personally.  If you disagree I guess I could let it slide, though.

Comment (mees-) [reply bin/cordlr]: The reason i did that is because i ran into an issue where when i would run cordlr setup before the package.json was created it would crash because bin/cordlr required the start method and that required the bin/config.js file and that file only works when a package.json is already present

Comment (mees-) [reply bin/cordlr]: This way the main file only requires the files it is actually going to need

File: bin/methods/setup.js [hunk 2]
```diff
@@ -3,22 +3,44 @@ const path = require('path')
...
+        const newPackage = Object.assign({}, existingPackage, { cordlr: config })
+        fs.writeFile(packagePath, JSON.stringify(newPackage), (err) => {
+          if (err) {
+            reject(err)
+          } else {
+            resolve(newPackage)
+          }
+        })
+      }
+    })
```
Comment (jamen) [inline bin/methods/setup.js]: I think this code could repeat less by just handling the data inside the `if` and not doing separate calls:

``` js
let newPackage = JSON.parse(data)
if (!err) {
  const existingPackage = JSON.parse(data)
  Object.assing(newPackage, existingPackage)
}

fs.write(packagePath, JSON.stringify(newPackage), ...)
```

File: bin/cordlr (hunk 1 above)
Comment (jamen) [reply bin/cordlr]: Hmm, could we fix that in `config.js` then?  Maybe just do an empty object if the file doesn't exist, instead of this.

File: bin/methods/setup.js (hunk 2 above)
Comment (mees-) [reply bin/methods/setup.js]: I agree and i will improve that tomorrow after i've had some sleep (europe)

File: bin/methods/setup.js [hunk 3]
```diff
@@ -3,22 +3,44 @@ const path = require('path')
...
-    const config = pixie.compile(TEMPLATE, result)
+    const config = Object.assign({}, configDefaults, { token: result.token, prefix: result.prefix })
     writeConfig(config)
   }).start()
 }
 
-function writeConfig (content) {
-  fs.writeFile(path.join(CWD, '.cordlrrc'), content, log('Configuration created'))
+function writeConfig (config) {
+  return new Promise((resolve, reject) => {
```
Comment (jamen) [inline bin/methods/setup.js]: I wouldn't even create a promise here, or handle the ending in any way:
1. `writeConfig` is not used by anything else than this file. (just puts some functionality under a more readable name)
2. `writeConfig` is the last function called in `cordlr setup` before program exits, so we wouldn't need to handle anything after it.
3. Each method inside `bin/methods/` is ran once before the program exits, so we wouldn't need to handle anything after `setup.js`

If we did need any of these in the future, adding a callback wouldn't be very hard at all.  Which leads me onto some side notes:

Right now, we don't have any details about use of promises, but it's worth noting that we don't use them anywhere in our code, so this would be the only time it appears instantiated.  

To be honest: I prefer a factory function + callback style.  I find it turns out just as clean!  Promises are stateful, honestly hard to maintain, have weird behavior in Node, and aren't optimized in V8 yet (last time I checked).  Not to be a stickler, because I used promises in lots of projects, I just don't like maintaining them!

Tl;dr: Just remove the promise, because we don't need to handle when it ends, and we don't use them much in our code style.

**Review by jamen — APPROVED** [review]
Looks good!

**Comment by jamen** [comment]
Just gonna pull this locally and test it for a bit, then I'll merge.

File: bin/config.js [hunk 4]
```diff
@@ -1,19 +1,11 @@
-const rc = require('rc')
+const fsproxy = require('fs-proxy')
 
-module.exports = rc('cordlr', {
```
Comment (jamen) [inline bin/config.js:3]: Hmm, one thing I have noticed is that `rc` allowed you to use command line options:

``` js
cordlr --token='...' --prefix='...'
```

Which I use heavily in development, and for just fast configuring (when you are lazy).

**Review by jamen — CHANGES_REQUESTED** [review]

File: bin/config.js (hunk 4 above)
Comment (mees-) [reply bin/config.js:3]: I assume you don't want the command line options to be written to the config file?

Comment (jamen) [reply bin/config.js:3]: Correct. 

You could have it just start out with the command line options, and if there is a config alternative, overwrite it?

Maybe using `Object.assign` or something.  Not entirely sure with how that all would work with `fsproxy`.

Comment (mees-) [reply bin/config.js:3]: Maybe using `fsproxy` is too limiting with our goals in mind, i think im going to (slightly) rewrite the `config.js` file to just read once. Shoud i include a write method in the module.exports to write the config to the file?

Comment (jamen) [reply bin/config.js:3]: Yeah sure, that sounds good.

Btw, for the reading, you can use `require(absolutePath)` just to skip `JSON.parse` and stuff.

File: bin/config.js [hunk 5]
```diff
@@ -1,19 +1,31 @@
...
+try {
+  config = require(configPath)
+} catch (e) {
+  // file doesn't exist
+  config = { cordlr: configDefaults }
+}
+
+// insert command line options
+const args = minimist(process.argv.slice(2))
+for (const key of Object.keys(args)) {
```
Comment (jamen) [inline bin/config.js]: Hmm,instead of looping, could we use `Object.assign` maybe?:

``` js
let config = minimist(process.argv.slice(2), {
  // defaults and stuff here ...
})

// load config:
try {
  Object.assign(config, require(configPath))
} catch (e) {  }

// ...
```

**Review by jamen — CHANGES_REQUESTED** [review]

**Comment by jamen** [comment]
Otherwise. changes look good.  :+1:

File: bin/config.js (hunk 5 above)
Comment (mees-) [reply bin/config.js]: The reason i didn't use `Object.assign` is because i wanted to filter out the `_` property `minimist` might attach, since we have no way of interpreting those settings (yet)

Comment (jamen) [reply bin/config.js]: Fair enough.  Maybe:

``` js
delete config._ 
```

After minimist part.

**Review by jamen — APPROVED** [review]
